src/Microsoft.DotNet.XHarness.WebAssembly/xunit-runner.cs

```diff
@@ -0,0 +1,354 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+using System.Linq;
+
+using Xunit;
+using Xunit.Abstractions;
+using Xunit.Sdk;
+using Xunit.ConsoleClient;
+
+namespace Microsoft.DotNet.XHarness.WebAssembly
+{
+    class MsgBus : IMessageBus
+    {
+        public MsgBus(IMessageSink messageSink)
+        {
+            Sink = messageSink;
+        }
+
+        IMessageSink Sink { get; set; }
+
+        public bool QueueMessage(IMessageSinkMessage message)
+        {
+            return Sink.OnMessage(message);
+        }
+
+        public void Dispose()
+        {
+        }
+    }
+
+    class Discoverer : XunitTestFrameworkDiscoverer
+    {
+        private readonly List<ITestClass> _testClasses;
+        private readonly IMessageSink _sink;
+        private readonly ITestFrameworkDiscoveryOptions _discoveryOptions;
+        private readonly IXunitTestCollectionFactory _testCollectionFactory;
+        int _nenumerated;
+
+        public Discoverer(IAssemblyInfo assemblyInfo,
+                          ISourceInformationProvider sourceProvider,
+                          IMessageSink diagnosticMessageSink,
+                          ITestFrameworkDiscoveryOptions discoveryOptions,
+                          IXunitTestCollectionFactory collectionFactory = null)
+            : base(assemblyInfo, sourceProvider, diagnosticMessageSink, collectionFactory)
+        {
+            _testCollectionFactory = collectionFactory;
+            _sink = diagnosticMessageSink;
+            _discoveryOptions = discoveryOptions ?? throw new ArgumentNullException(nameof(discoveryOptions));
+
+            _testClasses = new List<ITestClass>();
+            foreach (ITypeInfo type in AssemblyInfo.GetTypes(false).Where(IsValidTestClass))
+                _testClasses.Add(CreateTestClass(type));
+        }
+
+        protected override ITestClass CreateTestClass(ITypeInfo @class)
+        {
+            return new TestClass(_testCollectionFactory.Get(@class), @class);
+        }
+
+        public void Discover()
+        {
+            Console.WriteLine("Discovering tests...");
+            using (var messageBus = new MsgBus(_sink))
+            {
+                foreach (ITestClass testClass in _testClasses)
+                {
+                    if (testClass.Class.Name == "System.Threading.ThreadPools.Tests.ThreadPoolTests")
+                        // FIXME: This invokes the static ctor which creates threads
+                        continue;
+                    try
+                    {
+                        FindTestsForType(testClass, false, messageBus, _discoveryOptions);
+                    }
+                    catch (Exception ex)
+                    {
+                        Console.WriteLine("EX: " + ex);
```

Comment:
continue here? Or do we want to increase the _nenumerated even with an exception in FindTestsForType?

Reply:
`_nenumerated` is used to show the progress on discovering test classes so it's why it's increasing regardless the errors.